tools/buildsys/src/builder.rs

```diff
@@ -86,269 +85,355 @@ lazy_static! {
-        BuildType::Package => "package",
-        BuildType::Variant => "variant",
-    };
+        // Create a directory for tracking outputs before we move them into position.
+        let marker_dir = create_marker_dir(
+            &self.target_build_args.build_type(),
+            &self.artifact_name,
+            &self.common_build_args.arch.to_string(),
+            &self.state_dir,
+        )?;
 
-    // Our dockerfile is in the Twoliter tools directory.
-    let twoliter_tools_dir =
-        env::var(TOOLS_DIR).context(error::EnvironmentSnafu { var: TOOLS_DIR })?;
-    let dockerfile = PathBuf::from(twoliter_tools_dir).join("Dockerfile");
+        // Clean up any previous outputs we have tracked.
+        clean_build_files(&marker_dir, &self.artifacts_dir)?;
 
-    let mut build = format!(
-        "build . \
+        let mut build = format!(
+            "build {context} \
         --target {target} \
         --tag {tag} \
         --file {dockerfile}",
-        dockerfile = dockerfile.display(),
-        target = target,
-        tag = tag,
-    )
-    .split_string();
-
-    build.extend(build_args);
-    build.build_arg("SDK", sdk);
-    build.build_arg("TOOLCHAIN", toolchain);
-    build.build_arg("NOCACHE", nocache.to_string());
-    // Avoid using a cached layer from a concurrent build in another checkout.
-    build.build_arg("TOKEN", token);
-
-    let create = format!("create --name {} {} true", tag, tag).split_string();
-    let cp = format!("cp {}:/output/. {}", tag, build_dir.display()).split_string();
-    let rm = format!("rm --force {}", tag).split_string();
-    let rmi = format!("rmi --force {}", tag).split_string();
-
-    // Clean up the stopped container if it exists.
-    let _ = docker(&rm, Retry::No);
-
-    // Clean up the previous image if it exists.
-    let _ = docker(&rmi, Retry::No);
-
-    // Build the image, which builds the artifacts we want.
-    // Work around transient, known failure cases with Docker.
-    docker(
-        &build,
-        Retry::Yes {
-            attempts: DOCKER_BUILD_MAX_ATTEMPTS,
-            messages: &[
-                &*DOCKER_BUILD_FRONTEND_ERROR,
-                &*DOCKER_BUILD_DEAD_RECORD_ERROR,
-                &*UNEXPECTED_EOF_ERROR,
-                &*CREATEREPO_C_READ_HEADER_ERROR,
-            ],
-        },
-    )?;
-
-    // Create a stopped container so we can copy artifacts out.
-    docker(&create, Retry::No)?;
-
-    // Copy artifacts into our output directory.
-    docker(&cp, Retry::No)?;
-
-    // Clean up our stopped container after copying artifacts out.
-    docker(&rm, Retry::No)?;
-
-    // Clean up our image now that we're done.
-    docker(&rmi, Retry::No)?;
-
-    // Copy artifacts to the expected directory and write markers to track them.
-    copy_build_files(&build_dir, output_dir)?;
+            context = self.context.display(),
+            dockerfile = self.dockerfile.display(),
+            target = self.target,
+            tag = self.tag,
```

Comment:
nit: the spacing on these lines seems off?

Reply:
> nit: the spacing on these lines seems off?

Yeah fmt is no help here. I've now aligned the multiline string line starts with the variable list below it.